File: LeetCode/p102.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode *root) {
        vector<vector<int>> result;
        queue<TreeNode *> q;
        q.push(root);
        vector<int> t;
        while (!q.empty()) {
            t.clear();
            for (int i = q.size() - 1; i >= 0; --i) {
                TreeNode *this_node = q.front();
                q.pop();
                if (this_node != nullptr) {
                    t.push_back(this_node->val);
                    q.push(this_node->left);
                    q.push(this_node->right);
                }
            }
            if (!t.empty())
                result.push_back(t);
        }

        return result;
    }
};

File: LeetCode/p102.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
use std::collections::VecDeque;

impl Solution {
  pub fn level_order(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
    let mut ret = Vec::<Vec<i32>>::new();
    Self::helper(&root, &mut ret);
    ret
  }

  fn helper(root: &Option<Rc<RefCell<TreeNode>>>, v: &mut Vec<Vec<i32>>) {
    if root.is_none() {
      return;
    }
    let mut queue: VecDeque<Rc<RefCell<TreeNode>>> = VecDeque::from([root.as_ref().unwrap().clone()]);
    while !queue.is_empty() {
      let mut tv: Vec<i32> = Vec::new();
      for _ in 0..queue.len() {
        let rc = queue.pop_front().unwrap();
        tv.push(rc.borrow().val.clone());
        if let Some(rc) = &rc.borrow().left {
          queue.push_back(rc.clone());
        };
        if let Some(rc) = &rc.borrow().right {
          queue.push_back(rc.clone());
        };
      }
      v.push(tv);
    }
  }
}

